core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -305,4 +304,11 @@ class HadoopMapReduceCommitProtocol(
         logWarning(s"Exception while aborting ${taskContext.getTaskAttemptID}", e)
     }
   }
+
+  private def cleanStagingDir(jobContext: JobContext): Unit = {
+    val fs = stagingDir.getFileSystem(jobContext.getConfiguration)
+    if (fs.exists(stagingDir)) {
```

Comment:
This needs two RPC calls to remove the directory now, isn't it also a performance issue?

BTW, does `FileSystem` provide an API with "delete if exists" semantic?

Reply:
@cloud-fan Thanks for the review. 
The overhead of new added RPC call of readFile is quite small compared to the unnecessary deletion of the nonexistent file. Since In NameNode, read lock is shared and write lock is exclusive. 

And I think checking before existence is more of a client-side design, when we are running "hadoop fs -rm hdfs://ns/file", the existence is checked before real delete too.

In fact, I think it may be better to add a variable to indicate if the stagingDirectory is created so that we don't need the first RPC to check if it's existence. @cloud-fan What do you think?

Reply:
> This needs two RPC calls to remove the directory now, isn't it also a performance issue?
> 
> BTW, does `FileSystem` provide an API with "delete if exists" semantic?

I thin here we can use `deleteOnExit` like `SaveAsHiveFile`
```
fs.deleteOnExit(dirPath)
```

Reply:
One thing about this method is it's used when creating the path, notifying the FileSystem to delete when FileSystem is closed. A little different from the current use of `fs.delete(stagingDir)`.

Reply:
> One thing about this method is it's used when creating the path, notifying the FileSystem to delete when FileSystem is closed. A little different from the current use of `fs.delete(stagingDir)`.

I mean we can do like `SaveAsHiveFile`, mkdir first and let it delete when system exit. Since `deleteOnExit` will check file exit when deleting.

Reply:
I think the stagingDir is not created`HadoopMapReduceCommitProtocol` or its subclass, it's just noting the directory and the actual `OutputCommiter` will do the creation.

Unless we are asking `HadoopMapReduceCommitProtocol` and `SQLHadoopMapReduceCommitProtocol` to create the stagingDir at the following situation.
```
   * The staging directory of this write job. Spark uses it to deal with files with absolute output
   * path, or writing data into partitioned directory with dynamicPartitionOverwrite=true.
```

Reply:
please don't use delete on exit. that delete-on-shutdown operation is slow if called, and will delete what is there even if someone added a new file there after your operation. stick to using it for test teardown only.